### PR TITLE
Engineer Buildings

### PR DESCRIPTION
**Pull request by qualitykk**
Implements engineer buildings

Things which need to be looked into in the future:
- Building blueprint animations not being updated
- UI/Sound warnings when buildings are damaged
- Building damage particles
- Building gibs
- Sentry target aquire / shoot sounds
- Teleporter idle sound

### CODE REVIEW

**Comment by Retr0-01** [comment]
First of all, the diff is too large to view so I cannot possibly review your changes. Please cleanup your branch by squashing some commits.

Second, there are merge conflicts, resolve them too.

**Comment by qualitykk** [comment]
@Retr0-01 both of those should be addressed now

**Review by Retr0-01 — CHANGES_REQUESTED** [review]
- When creating a dispenser, it creates a particle on its origin which is infinite and undestroyable.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/c0eca291-8aa4-49bb-b33a-925969ca9c34

- Your player is effectively stuck if you equip the building pad and have no metal to create a building. You cannot switch back to your weapons directly and pressing "Q" opens up the menu instead of closing the pad menu.
- While standing on top of a recharging teleporter (entrance), you won't be teleported when the charge is ready unless you move a little bit, in any direction.
- Remove leftover debug message:
```
Tried to build even though we cant, ignoring
   at TFS2.Builder.Build() in C:\Program Files (x86)\Steam\steamapps\common\sbox\addons\TF-Source-2\code\Player\Weapons\Builder.cs:line 50
   at TFS2.Builder.Attack() in C:\Program Files (x86)\Steam\steamapps\common\sbox\addons\TF-Source-2\code\Player\Weapons\Builder.cs:line 36
```
- You can build through walls.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/043d9472-e334-4acc-8325-9eff311d78ff

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/ceac7b40-1c76-42dc-b471-99e982827937

- Your sentry might eventually turn against you, try to kill you, and if it succeeds it will kill itself after that.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/1dc45125-dbdd-41b6-8992-ec029c4b9b6c

- Adding to the above issue, I cannot destroy my teleporter on the above clip, it's like I suddenly do not own these buildings anymore.

File: code/Player/Buildings.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	{
+		int given = (int)MathF.Min(MaxMetal - Metal,amount);
+		Metal += amount;
+		if ( Metal > MaxMetal )
+			Metal = MaxMetal;
+
+		return given;
+	}
+
+	public void DestroyBuildings()
```
Comment (Retr0-01) [inline code/Player/Buildings.cs:105]: Sometimes when the round or the game restarts (mp_restartround and mp_restartgame) your sentry gun and/or teleporter entrance/exit will not be destroyed.

File: code/Player/Buildings/TeleporterEntrance.cs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	public override void Tick()
+	{
+		if ( IsCarried || !IsInitialized ) return;
+
+		base.Tick();
+
+		if( IsPaired )
+		{
+			// Rotation between this and the other teleporter
+			var direction = Rotation.LookAt( LinkedTeleporter.Position - Position );
```
Comment (Retr0-01) [inline code/Player/Buildings/TeleporterEntrance.cs:82]: NRE occurring if a tele pair that is being constructed right now gets destroyed by round/game restart.
Not persistent.
```
Error calling event 'server.tick' on 'TeleporterEntrance 438'
System.NullReferenceException: Object reference not set to an instance of an object.
   at TFS2.TeleporterEntrance.Tick() in C:\Program Files (x86)\Steam\steamapps\common\sbox\addons\TF-Source-2\code\Player\Buildings\TeleporterEntrance.cs:line 82
   at InvokeStub_TeleporterEntrance.Tick(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

**Review by Retr0-01 — CHANGES_REQUESTED** [review]
- Buildings should be solid against their owner. Currently the owner engineer can pass through them.
- Dispenser particle will stay alive if the dispenser is destroyed while healing the player.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/3a7fbea6-81ea-478b-ae8d-86c0606137bb

- The first rocket of a lvl 3 sentry gun will always miss-fire and hit nothing. Notice how the rocket is fired immediately when a target is spotted on TF:S2 while on TF2 it waits and then fires along with the bullets, hitting the target.

TF:S2:

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/b8de2ea6-bc09-4d1b-8720-08beda4162f9

Live TF2:

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/e6002bc9-30a9-4857-bd38-b46a480960e8

- The sentry gun upgrade animations are cut and not finished properly.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/b92c8189-5efb-4cfd-a5f6-e4f9cb9c8827

- The check bounds for placing a teleporter entrance/exit need to be increased. Players can easily get stuck in walls if a teleporter is placed correctly.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/9178edb3-cb47-4252-b406-0bddaac6d08d

File: data/weapons/destroy_pda.tfweapon [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "Title": "Destroy PDA",
-  "EngineClass": "",
+  "EngineClass": "tf_weapon_destruction_pda",
   "Hidden": false,
   "WorldModel": "models/weapons/c_models/c_pda_engineer/c_pda_engineer.vmdl",
   "Owners": {
```
Comment (badandbest) [inline data/weapons/destroy_pda.tfweapon:6]: PDA2 is not an animation. It should be PDA.

**Review by badandbest — CHANGES_REQUESTED** [review]
- Buildings can't be upgraded when crouching.

https://github.com/AmperSoftware/TF-Source-2/assets/91832803/296e2431-28a8-4192-acd5-04df0fb4a718

- Teleport Exit's level does not increase.

https://github.com/AmperSoftware/TF-Source-2/assets/91832803/3225bca0-ea46-497a-8264-44ab35240daf

- Placing down two teleporters without speeding up construction will make them take 1 metal before upgrading
  - This means you'll be 1 metal short from upgrading to level 2.

https://github.com/AmperSoftware/TF-Source-2/assets/91832803/417071cc-49ec-498e-9b04-f503fe8fc997

- Picking up an exit when you're about to teleport will make you teleport to it once you place it down again.

https://github.com/AmperSoftware/TF-Source-2/assets/91832803/43c1a524-5d8e-4b19-845b-01e81d037de5

- Teleporter recharge bar does not stretch to the end.

https://github.com/AmperSoftware/TF-Source-2/assets/91832803/b9ab40b3-53d6-43be-9934-8b6351e6b5e5

**Comment by justyn0** [comment]
Something to note, It seems like the engineer could go through his buildings. However, in Live TF2 the Engineer is the only class that could collide with his own buildings. Other classes or a friendly Engineer could pass through every building except for the Teleporters. Since the player steps on the teleporter and wait to be transported

**Comment by qualitykk** [comment]
Those issues should be fixed now.
The only thing i cant fix is making it so the buildings collide with the player who built them, right now you cant do per-entity collision checks in sbox.

**Comment by bodolaz146** [comment]
I think the real problem lies with this collision check for the players:
```csharp
if ( !tf_collide_with_teammates )
    tr = tr.WithoutTags( playerTeam.GetTag() );
```
Essentially, this check means "we should never collide with our team things". Obviously, it's not always true (ex. our buildings, our payload cart, or even the fact that we can move our teammates in tf2). This check can be changed to "we should always collide with enemy team things":
```csharp
tr = tr.WithAnyTags( playerTeam.GeEnemytTag() );
```
This way we can add a special tag for our buildings like this:
```csharp
Tags.Add( $"building_{Owner.Client.Id}" );
```
And include a check for player collision:
```csharp
if ( Player.Buildings.Count != 0 )
{
    tr = tr.WithAnyTags( $"building_{Player.Client.Id}" );
}
```
Unfortunately, this will break a lot of collision-related code. The "solid" tag is slapped on every entity possible, and we will have to change it on a case-by-case basis. For example, for buildings, we can use the tag "building" instead and add it to projectile collision checks. 
Implementing this approach may take some time, but it should be feasible.

**Review by Retr0-01 — APPROVED** [review]
Buildings seem stable enough now to be released. Good job.

PS: Upgrade animations for the sentry are still kinda messed up but I won't hold this PR for such a small issue.

https://github.com/AmperSoftware/TF-Source-2/assets/61121754/ba0efc3d-0995-4737-b621-6ff20253891d

**Pull request merged**